README.md
# LLRegex

[![Build Status](https://travis-ci.org/LittleRockInGitHub/LLRegex.svg?branch=master)](https://travis-ci.org/LittleRockInGitHub/LLRegex)
[![CocoaPods Compatible](https://img.shields.io/cocoapods/v/LLRegex.svg)](https://img.shields.io/cocoapods/v/LLRegex.svg)
[![Platform](https://img.shields.io/cocoapods/p/LLRegex.svg)](https://img.shields.io/cocoapods/p/LLRegex.svg)

Regular expression library in Swift, wrapping NSRegularExpression.
Don't hesitate to try out on [playground](https://github.com/LittleRockInGitHub/LLRegex/blob/master/LLRegex.playground.zip).

## Features
 * Value Semantics
 * Enumerates matches with Sequence
 * Named capture group (unavailable on iOS 8)
 * Range supported (NSRange eliminated)
 * Regex Options, Match Options
 * Find & Replace with flexibility
 * String matching, replacing, splitting

## Communication
* If you **found a bug**, open an issue, typically with related pattern.
* If you **have a feature request**, open an issue.
 
## Requirements

- iOS 9.0+ / macOS 10.10+ / tvOS 9.0+ / watchOS 2.0+
- Xcode 11.3+
- Swift 5.0+

## Installation

### CocoaPods

```ruby
pod 'LLRegex', '~> 1.4'
```

### Swift Package Manager

```swift
dependencies: [
    .Package(url: "https://github.com/LittleRockInGitHub/LLRegex.git", majorVersion: 1)
]
```

## Usage

### Making a Regex  

```swift
let numbers = Regex("(\\d)(\\d+)(\\d)")

let insensitive = Regex("LLRegex", options: [.caseInsensitive])

let runtimeError = Regex("")    // Runtime error would be raised

let invalid = try? Regex(pattern: "")   // nil returned
```

### Searching
 Method `matches(in:options:range:)` returns a sequence producing matches **lazily**, which is the only one for searching. All other variants were dropped thanks to the power of Sequence.
 
```swift
let s = "123-45-6789-0-123-45-6789-01234"
let subrange = s.characters.dropFirst(3).startIndex..<s.endIndex

for match in numbers.matches(in: s) {
    // enumerating
    match.matched
}

if let first = numbers.matches(in: s).first {
    // first match
    first.matched
}

let allMatches: [Match] = numbers.matches(in: s).all // all matches

let subrangeMatches = numbers.matches(in: s, options: [.withTransparentBounds], range: subrange)

for case let match in subrangeMatches.dropFirst(1) where match.matched != "6789" {
    match.matched
}
```

### Match & Capture Groups

```swift
if let first = numbers.matches(in: s).first {
    
    first.matched
    first.range
    first.groups.count
    first.groups[1].matched
    first.groups[1].range
    
    let replacement = first.replacement(withTemplate: "$3$2$1")     // Replacement with template
}
```
  
### Named Capture Group (unavailable on iOS 8)
Named capture group feature is enabled when `.namedCaptureGroups` is set.

```swift
let named = Regex("(?<year>\\d+)-(?<month>\\d+)-(?<day>\\d+)", options: .namedCaptureGroups)
let s = "Today is 2017-06-23."

for m in named.matches(in: s) {
    m.groups["year"]?.matched
}

named.replacingAllMatches(in: s, replacement: .replaceWithTemplate("${month}/${day}/${year}")) // Today is 06/23/2017.
```  
- Note: If the comment in pattern contains the notation of capture group, the detection for named capture group will fail.

### Replacing

```swift
numbers.replacingFirstMatch(in: s, replacement: .remove)

numbers.replacingAllMatches(in: s, range: subrange, replacement: .replaceWithTemplate("$3$2$1"))
```

Flexible Find & Replace is offered by `replacingMatches(in:options:range:replacing:)`.

```swift  
numbers.replacingMatches(in: s) { (idx, match) -> Match.Replacing in
    
    switch idx {
    case 0:
        return .keep    // Keeps unchanged
    case 1:
        return .remove  // Removes the matched string
    case 2:
        return .replaceWithTemplate("($1-$3)")    // Replaces with template
    case 3:
        return .replaceWithString(String(match.matched.characters.reversed()))   // Replaces with string
    default:
        return .stop    // Stops replacing
    }
}
```

### String Matching

```swift
"123".isMatching("\\d+")
"llregex".isMatching(insensitive)   // Regex is accepted
"123-456".isMatching("\\d+")    // isMatching(_:) checks whether matches entirely
```

### String Replacing
- Note: Two variants - one with pattern and template label, the other is not.

```swift  
"123".replacingAll("1", with: "$0")
"123-321".replacingAll(pattern: "(\\d)(\\d)", withTemplate: "$2$1")

s.replacingFirst(pattern: numbers, in: subrange, withTemplate: "!")
```

### String Splitting

```swift
s.split(seperator: "\\d")
s.split(seperator: numbers, maxSplits: 2, omittingEmptyString: false)
```



CHANGELOG.md
# Change Log

## 1.4.0
- Compatible with Swift 5.0

## 1.3.0
- Compatible with Swift 4.0
- `CaptureGroup` conforms to `MatchProtocol`
- `matched` on `CaptureGroup` became Non-optional. Empty string is returned if conversion is failed.
- `range` on `Match` became Optional. `nil` is returned if conversion from NSRange is failed.
- `String.split` returns `[Substring]` instead of `[String]`
- Performance improved

## 1.2.1
- Compatible with Swift 3.2

## 1.2.0
- Named capture group
- `Regex.Options` is not type alias of `NSRegularExpession.Options` any longer.
- Add `Regex.Options.namedCaputreGroups`
- `Match.Options` is not type alias of `NSRegularExpession.MatchingOptions` any longer.
- Add `Match.Options.reportProgress` & `Match.Options.reportCompletion`
- 
## 1.1.1
- Compatible with Swift 3.2
- Swift Package Manager supported

## 1.1.0
- `Regex.init`` changed - one accepts string literal, the other one accepts string may throwing error.
- macOS, tvOS, watchOS supported.

## 1.0.0
- LLRegex initial version


Project/LLRegex/Playground.swift
//
//  Playground.swift
//  LLRegex
//
//  Created by Rock Yang on 2017/6/10.
//  Copyright © 2017年 Rock Young. All rights reserved.
//

import Foundation

let numbers = Regex("(\\d)(\\d+)(\\d)")

let insensitive = Regex("LLRegex", options: [.caseInsensitive])

let runtimeError = Regex("")    // Runtime error would be raised

let invalid = try? Regex(pattern: "")   // nil returned

let s = "123-45-6789-0-123-45-6789-01234"

let subrange = s.characters.dropFirst(3).startIndex..<s.endIndex

struct Playground {
    
    public func search() {
        
        for match in numbers.matches(in: s) {
            // enumerating
        }
        
        if let first = numbers.matches(in: s).first {
            // first match
        }
        
        let allMatches: [Match] = numbers.matches(in: s).all // all matches
        
        let subrangeMatches = numbers.matches(in: s, options: [.withTransparentBounds], range: subrange)
        
        for case let match in subrangeMatches.dropFirst(1) where match.matched != "45" {
            
        }
    }
    
    public func match() {
        
        if let first = numbers.matches(in: s).first {
            
            first.matched
            first.range
            first.groups.count
            first.groups[1].matched
            first.groups[1].range
            
            let replacement = first.replacement(withTemplate: "$3$2$1")
            
        }
    }
    
    
    public func namedCaptureGroups() {
    
        let named = Regex("(?<year>\\d+)-(?<month>\\d+)-(?<day>\\d+)", options: .namedCaptureGroups)
        let s = "Today is 2017-06-23."
        
        for m in named.matches(in: s) {
            m.groups["year"]?.matched
        }
        
        named.replacingAllMatches(in: s, replacement: .replaceWithTemplate("${month}/${day}/${year}"))
    }
    
    public func replace() {
        
        numbers.replacingFirstMatch(in: s, replacement: .remove)
        
        numbers.replacingAllMatches(in: s, range: subrange, replacement: .replaceWithTemplate("$3$2$1"))
        
        numbers.replacingMatches(in: s) { (idx, match) -> Match.Replacing in
            
            switch idx {
            case 0:
                return .keep    // Keep unchanged
            case 1:
                return .remove  // Remove the matched string
            case 2:
                return .replaceWithTemplate("$1-$3")    // Replace with template
            case 3:
                return .replaceWithString({
                    return String(match.matched.characters.reversed())  // Relace with string
                    }())
            default:
                return .stop    // Stop replacing
            }
        }
    }
    
    public func string() {
        
        "123".isMatching("\\d+")
        "llregex".isMatching(insensitive)
        "123-456".isMatching("\\d+")
        
        "123".replacingAll("1", with: "!")
        "123".replacingAll(pattern: "(\\d)(\\d)", withTemplate: "$2$1")
        
        "123".replacingFirst(pattern: numbers, in: subrange, withTemplate: "!")
        
        s.split(seperator: "\\d")
    }
    
    
}
